tensornetwork/tensornetwork.py

```diff
@@ -317,8 +318,26 @@ def update_axis(self, old_axis: int, old_node: Node, new_axis: int,
                        "node1: '{}', axis1: {}, node2: '{}', axis2: {}".format(
                            self, old_node, old_axis, self.node1, self.axis1,
                            self.node2, self.axis2))
-
-  def is_dangling(self):
+  @property
+  def node1(self) -> Node:
+    val = self._node1()
+    if val is None:
+      raise ValueError("node1 for edge '{}' no longer exists.".format(self))
+    return val
+
+  @property
+  def node2(self) -> Optional[Node]:
+    return self._node2() if self._node2 else None  
+  
+  @node1.setter
+  def node1(self, node: Node) -> None:
+    self._node1 = weakref.ref(node)
+
+  @node2.setter
+  def node2(self, node: Node) -> None:
+    self._node2 = weakref.ref(node) if node else None
```

Comment:
We probably don't need to check `node` here, so the conditional can be removed.

Reply:
There can't exist weak references to NoneType. Keeping the conditional for now.

Reply:
Yes, but `node` is of type `Node`, so cannot be `None`. Unless the type annotation is incorrect...

Reply:
It was incorrect. surprised pytype didn't catch it...